{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "736a2faf",
   "metadata": {},
   "source": [
    "# Example notebook\n",
    "\n",
    "This notebook can be used for testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb4db0c2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/scratch/project_2007145/antkark/MMB-117/MMB-117_EnvironmentalMicrobiology'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "895413cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 8\r\n",
      "-rw-rw---- 1 antkark project_2007145 1091 Jan 19 10:15 Examples.ipynb\r\n",
      "-rw-rw---- 1 antkark project_2007145  749 Jan 19 10:15 README.md\r\n"
     ]
    }
   ],
   "source": [
    "ls -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ce51b6b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "            FastQC - A high throughput sequence QC analysis tool\r\n",
      "\r\n",
      "SYNOPSIS\r\n",
      "\r\n",
      "\tfastqc seqfile1 seqfile2 .. seqfileN\r\n",
      "\r\n",
      "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \r\n",
      "           [-c contaminant file] seqfile1 .. seqfileN\r\n",
      "\r\n",
      "DESCRIPTION\r\n",
      "\r\n",
      "    FastQC reads a set of sequence files and produces from each one a quality\r\n",
      "    control report consisting of a number of different modules, each one of \r\n",
      "    which will help to identify a different potential type of problem in your\r\n",
      "    data.\r\n",
      "    \r\n",
      "    If no files to process are specified on the command line then the program\r\n",
      "    will start as an interactive graphical application.  If files are provided\r\n",
      "    on the command line then the program will run with no user interaction\r\n",
      "    required.  In this mode it is suitable for inclusion into a standardised\r\n",
      "    analysis pipeline.\r\n",
      "    \r\n",
      "    The options for the program as as follows:\r\n",
      "    \r\n",
      "    -h --help       Print this help file and exit\r\n",
      "    \r\n",
      "    -v --version    Print the version of the program and exit\r\n",
      "    \r\n",
      "    -o --outdir     Create all output files in the specified output directory.\r\n",
      "                    Please note that this directory must exist as the program\r\n",
      "                    will not create it.  If this option is not set then the \r\n",
      "                    output file for each sequence file is created in the same\r\n",
      "                    directory as the sequence file which was processed.\r\n",
      "                    \r\n",
      "    --casava        Files come from raw casava output. Files in the same sample\r\n",
      "                    group (differing only by the group number) will be analysed\r\n",
      "                    as a set rather than individually. Sequences with the filter\r\n",
      "                    flag set in the header will be excluded from the analysis.\r\n",
      "                    Files must have the same names given to them by casava\r\n",
      "                    (including being gzipped and ending with .gz) otherwise they\r\n",
      "                    won't be grouped together correctly.\r\n",
      "                    \r\n",
      "    --nano          Files come from nanopore sequences and are in fast5 format. In\r\n",
      "                    this mode you can pass in directories to process and the program\r\n",
      "                    will take in all fast5 files within those directories and produce\r\n",
      "                    a single output file from the sequences found in all files.                    \r\n",
      "                    \r\n",
      "    --nofilter      If running with --casava then don't remove read flagged by\r\n",
      "                    casava as poor quality when performing the QC analysis.\r\n",
      "                   \r\n",
      "    --extract       If set then the zipped output file will be uncompressed in\r\n",
      "                    the same directory after it has been created.  By default\r\n",
      "                    this option will be set if fastqc is run in non-interactive\r\n",
      "                    mode.\r\n",
      "                    \r\n",
      "    -j --java       Provides the full path to the java binary you want to use to\r\n",
      "                    launch fastqc. If not supplied then java is assumed to be in\r\n",
      "                    your path.\r\n",
      "                   \r\n",
      "    --noextract     Do not uncompress the output file after creating it.  You\r\n",
      "                    should set this option if you do not wish to uncompress\r\n",
      "                    the output when running in non-interactive mode.\r\n",
      "                    \r\n",
      "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\r\n",
      "                    show data for every base in the read.  WARNING: Using this\r\n",
      "                    option will cause fastqc to crash and burn if you use it on\r\n",
      "                    really long reads, and your plots may end up a ridiculous size.\r\n",
      "                    You have been warned!\r\n",
      "                    \r\n",
      "    --min_length    Sets an artificial lower limit on the length of the sequence\r\n",
      "                    to be shown in the report.  As long as you set this to a value\r\n",
      "                    greater or equal to your longest read length then this will be\r\n",
      "                    the sequence length used to create your read groups.  This can\r\n",
      "                    be useful for making directly comaparable statistics from \r\n",
      "                    datasets with somewhat variable read lengths.\r\n",
      "                    \r\n",
      "    -f --format     Bypasses the normal sequence file format detection and\r\n",
      "                    forces the program to use the specified format.  Valid\r\n",
      "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\r\n",
      "                    \r\n",
      "    -t --threads    Specifies the number of files which can be processed\r\n",
      "                    simultaneously.  Each thread will be allocated 250MB of\r\n",
      "                    memory so you shouldn't run more threads than your\r\n",
      "                    available memory will cope with, and not more than\r\n",
      "                    6 threads on a 32 bit machine\r\n",
      "                  \r\n",
      "    -c              Specifies a non-default file which contains the list of\r\n",
      "    --contaminants  contaminants to screen overrepresented sequences against.\r\n",
      "                    The file must contain sets of named contaminants in the\r\n",
      "                    form name[tab]sequence.  Lines prefixed with a hash will\r\n",
      "                    be ignored.\r\n",
      "\r\n",
      "    -a              Specifies a non-default file which contains the list of\r\n",
      "    --adapters      adapter sequences which will be explicity searched against\r\n",
      "                    the library. The file must contain sets of named adapters\r\n",
      "                    in the form name[tab]sequence.  Lines prefixed with a hash\r\n",
      "                    will be ignored.\r\n",
      "                    \r\n",
      "    -l              Specifies a non-default file which contains a set of criteria\r\n",
      "    --limits        which will be used to determine the warn/error limits for the\r\n",
      "                    various modules.  This file can also be used to selectively \r\n",
      "                    remove some modules from the output all together.  The format\r\n",
      "                    needs to mirror the default limits.txt file found in the\r\n",
      "                    Configuration folder.\r\n",
      "                    \r\n",
      "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\r\n",
      "                    module. Specified Kmer length must be between 2 and 10. Default\r\n",
      "                    length is 7 if not specified.\r\n",
      "                    \r\n",
      "   -q --quiet       Supress all progress messages on stdout and only report errors.\r\n",
      "   \r\n",
      "   -d --dir         Selects a directory to be used for temporary files written when\r\n",
      "                    generating report images. Defaults to system temp directory if\r\n",
      "                    not specified.\r\n",
      "                    \r\n",
      "BUGS\r\n",
      "\r\n",
      "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\r\n",
      "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\r\n",
      "                   \r\n",
      "    \r\n"
     ]
    }
   ],
   "source": [
    "!fastqc -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cb409465",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                \n",
      " \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=404620;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ ğŸ” \u001b[2m| v1.\u001b[0m\u001b[1;2;36m14\u001b[0m                                                         \n",
      "\u001b[1m                                                                                \u001b[0m\n",
      "\u001b[1m \u001b[0m\u001b[1;33mUsage: \u001b[0m\u001b[1mmultiqc [\u001b[0m\u001b[1;36mOPTIONS\u001b[0m\u001b[1m] [\u001b[0m\u001b[1;36mANALYSIS\u001b[0m\u001b[1m \u001b[0m\u001b[1;36mDIRECTORY\u001b[0m\u001b[1m]\u001b[0m\u001b[1m                                 \u001b[0m\u001b[1m \u001b[0m\n",
      "\u001b[1m                                                                                \u001b[0m\n",
      " MultiQC aggregates results from bioinformatics analyses across many samples    \n",
      " into a single report.                                                          \n",
      " \u001b[2mIt searches a given directory for analysis logs and compiles a \u001b[0m\u001b[1;2;36mHTML\u001b[0m\u001b[2m report. \u001b[0m   \n",
      " \u001b[2mIt's a general use tool, perfect for summarising the output from numerous \u001b[0m     \n",
      " \u001b[2mbioinformatics tools.\u001b[0m                                                          \n",
      " \u001b[2mTo run, supply with one or more directory to scan for analysis results. For \u001b[0m   \n",
      " \u001b[2mexample, to run in the current working directory, use '\u001b[0m\u001b[1;2;34mmultiqc .\u001b[0m\u001b[2m'\u001b[0m              \n",
      "                                                                                \n",
      "\u001b[2mâ•­â”€\u001b[0m\u001b[2m Main options \u001b[0m\u001b[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\u001b[0m\u001b[2mâ”€â•®\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-force\u001b[0m            \u001b[1;32m-f\u001b[0m  Overwrite any existing reports                        \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-config\u001b[0m           \u001b[1;32m-c\u001b[0m  Specific config file to load, after those in MultiQC  \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        dir / home dir / working dir.                         \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-cl\u001b[0m\u001b[1;36m-config\u001b[0m            Specify MultiQC config \u001b[1;36mYAML\u001b[0m on the command line       \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(TEXT)                                         \u001b[0m       \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-filename\u001b[0m         \u001b[1;32m-n\u001b[0m  Report filename. Use '\u001b[3;33mstdout\u001b[0m' to print to standard    \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        out.                                                  \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-outdir\u001b[0m           \u001b[1;32m-o\u001b[0m  Create report in the specified output directory.      \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(TEXT)                                          \u001b[0m      \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ignore\u001b[0m           \u001b[1;32m-x\u001b[0m  Ignore analysis files \u001b[2;33m(GLOB EXPRESSION)\u001b[0m               \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ignore\u001b[0m\u001b[1;36m-samples\u001b[0m       Ignore sample names \u001b[2;33m(GLOB EXPRESSION)\u001b[0m                 \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ignore\u001b[0m\u001b[1;36m-symlinks\u001b[0m      Ignore symlinked directories and files                \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-file\u001b[0m\u001b[1;36m-list\u001b[0m        \u001b[1;32m-l\u001b[0m  Supply a file containing a list of file paths to be   \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        searched, one per row                                 \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯\u001b[0m\n",
      "\u001b[2mâ•­â”€\u001b[0m\u001b[2m Choosing modules to run \u001b[0m\u001b[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\u001b[0m\u001b[2mâ”€â•®\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-module\u001b[0m     \u001b[1;32m-m\u001b[0m  Use only this module. Can specify multiple times.           \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                  \u001b[2;33m(MODULE NAME)                                    \u001b[0m           \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-exclude\u001b[0m    \u001b[1;32m-e\u001b[0m  Do not use this module. Can specify multiple times.         \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                  \u001b[2;33m(MODULE NAME)                                      \u001b[0m         \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-tag\u001b[0m            Use only modules which tagged with this keyword \u001b[2;33m(TEXT)\u001b[0m      \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-view\u001b[0m\u001b[1;36m-tags\u001b[0m      View the available tags and which modules they load         \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯\u001b[0m\n",
      "\u001b[2mâ•­â”€\u001b[0m\u001b[2m Sample handling \u001b[0m\u001b[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\u001b[0m\u001b[2mâ”€â•®\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-dirs\u001b[0m           \u001b[1;32m-d\u001b[0m   Prepend directory to sample names                      \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-dirs\u001b[0m\u001b[1;36m-depth\u001b[0m     \u001b[1;32m-dd\u001b[0m  Prepend \u001b[3;33mn\u001b[0m directories to sample names. Negative number \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                       to take from start of path.                            \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                       \u001b[2;33m(INTEGER)                                             \u001b[0m \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-fullnames\u001b[0m      \u001b[1;32m-s\u001b[0m   Do not clean the sample names \u001b[3m(leave as full file \u001b[0m     \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                       \u001b[3mname)\u001b[0m                                                  \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-fn_as_s_name\u001b[0m        Use the log filename as the sample name                \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-replace\u001b[0m\u001b[1;36m-names\u001b[0m       \u001b[1;36mTSV\u001b[0m file to rename sample names during report          \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                       generation                                             \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                       \u001b[2;33m(PATH)                                                \u001b[0m \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯\u001b[0m\n",
      "\u001b[2mâ•­â”€\u001b[0m\u001b[2m Report customisation \u001b[0m\u001b[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\u001b[0m\u001b[2mâ”€â•®\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-title\u001b[0m            \u001b[1;32m-i\u001b[0m  Report title. Printed as page header, used for        \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        filename if not otherwise specified.                  \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-comment\u001b[0m          \u001b[1;32m-b\u001b[0m  Custom comment, will be printed at the top of the     \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        report.                                               \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-template\u001b[0m         \u001b[1;32m-t\u001b[0m  Report template to use.                               \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(default|default_dev|gathered|geo|sections|simple)\u001b[0m    \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-sample\u001b[0m\u001b[1;36m-names\u001b[0m         \u001b[1;36mTSV\u001b[0m file containing alternative sample names for      \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        renaming buttons in the report                        \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-sample\u001b[0m\u001b[1;36m-filters\u001b[0m       \u001b[1;36mTSV\u001b[0m file containing show/hide patterns for the report \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-custom\u001b[0m\u001b[1;36m-css-file\u001b[0m      Custom \u001b[1;36mCSS\u001b[0m file to add to the final report \u001b[2;33m(PATH)\u001b[0m     \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯\u001b[0m\n",
      "\u001b[2mâ•­â”€\u001b[0m\u001b[2m Output files \u001b[0m\u001b[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\u001b[0m\u001b[2mâ”€â•®\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-flat\u001b[0m          \u001b[1;32m-fp\u001b[0m  Use only flat plots \u001b[3m(static images)\u001b[0m                     \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-interactive\u001b[0m   \u001b[1;32m-ip\u001b[0m  Use only interactive plots \u001b[3m(in-browser Javascript)\u001b[0m      \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-export\u001b[0m        \u001b[1;32m-p\u001b[0m   Export plots as static images in addition to the report \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-data\u001b[0m\u001b[1;36m-dir\u001b[0m           Force the parsed data directory to be created.          \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-no\u001b[0m\u001b[1;36m-data-dir\u001b[0m        Prevent the parsed data directory from being created.   \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-data\u001b[0m\u001b[1;36m-format\u001b[0m   \u001b[1;32m-k\u001b[0m   Output parsed data in a different format.               \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                      \u001b[2;33m(tsv|json|yaml)                          \u001b[0m               \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-zip\u001b[0m\u001b[1;36m-data-dir\u001b[0m  \u001b[1;32m-z\u001b[0m   Compress the data directory.                            \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-no\u001b[0m\u001b[1;36m-report\u001b[0m          Do not generate a report, only export data and plots    \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-pdf\u001b[0m                Creates \u001b[1;36mPDF\u001b[0m report with the \u001b[3m'simple'\u001b[0m template. Requires \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                      \u001b]8;id=135844;https://pandoc.org/\u001b\\Pandoc\u001b]8;;\u001b\\ to be installed.                                 \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯\u001b[0m\n",
      "\u001b[2mâ•­â”€\u001b[0m\u001b[2m MultiQC behaviour \u001b[0m\u001b[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\u001b[0m\u001b[2mâ”€â•®\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-verbose\u001b[0m           \u001b[1;32m-v\u001b[0m  Increase output verbosity. \u001b[2;33m(INTEGER RANGE)\u001b[0m           \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-quiet\u001b[0m             \u001b[1;32m-q\u001b[0m  Only show log warnings                               \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-lint\u001b[0m                  Use strict linting (validation) to help code         \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                         development                                          \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-profile\u001b[0m\u001b[1;36m-runtime\u001b[0m       Add analysis of how long MultiQC takes to run to the \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                         report                                               \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-no\u001b[0m\u001b[1;36m-megaqc-upload\u001b[0m      Don't upload generated report to MegaQC, even if     \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m                         MegaQC options are found                             \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-no\u001b[0m\u001b[1;36m-ansi\u001b[0m               Disable coloured log output                          \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-version\u001b[0m               Show the version and exit.                           \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ”‚\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-help\u001b[0m              \u001b[1;32m-h\u001b[0m  Show this message and exit.                          \u001b[2mâ”‚\u001b[0m\n",
      "\u001b[2mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯\u001b[0m\n",
      "                                                                                \n",
      " See \u001b]8;id=746092;http://multiqc.info\u001b\\http://multiqc.info\u001b]8;;\u001b\\ for more details.                                      \n",
      "\n"
     ]
    }
   ],
   "source": [
    "!multiqc -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1a8225c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt version 4.2\r\n",
      "\r\n",
      "Copyright (C) 2010-2022 Marcel Martin <marcel.martin@scilifelab.se>\r\n",
      "\r\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\r\n",
      "\r\n",
      "Usage:\r\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\r\n",
      "\r\n",
      "For paired-end reads:\r\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\r\n",
      "\r\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\r\n",
      "characters are supported. All reads from input.fastq will be written to\r\n",
      "output.fastq with the adapter sequence removed. Adapter matching is\r\n",
      "error-tolerant. Multiple adapter sequences can be given (use further -a\r\n",
      "options), but only the best-matching adapter will be removed.\r\n",
      "\r\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\r\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\r\n",
      "standard input/output. Without the -o option, output is sent to standard output.\r\n",
      "\r\n",
      "Citation:\r\n",
      "\r\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\r\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\r\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\r\n",
      "\r\n",
      "Run \"cutadapt --help\" to see all command-line options.\r\n",
      "See https://cutadapt.readthedocs.io/ for full documentation.\r\n",
      "\r\n",
      "Options:\r\n",
      "  -h, --help            Show this help message and exit\r\n",
      "  --version             Show version number and exit\r\n",
      "  --debug               Print debug log. Use twice to also print DP matrices\r\n",
      "  -j CORES, --cores CORES\r\n",
      "                        Number of CPU cores to use. Use 0 to auto-detect.\r\n",
      "                        Default: 1\r\n",
      "\r\n",
      "Finding adapters:\r\n",
      "  Parameters -a, -g, -b specify adapters to be removed from each read (or from\r\n",
      "  R1 if data is paired-end. If specified multiple times, only the best\r\n",
      "  matching adapter is trimmed (but see the --times option). Use notation\r\n",
      "  'file:FILE' to read adapter sequences from a FASTA file.\r\n",
      "\r\n",
      "  -a ADAPTER, --adapter ADAPTER\r\n",
      "                        Sequence of an adapter ligated to the 3' end (paired\r\n",
      "                        data: of the first read). The adapter and subsequent\r\n",
      "                        bases are trimmed. If a '$' character is appended\r\n",
      "                        ('anchoring'), the adapter is only found if it is a\r\n",
      "                        suffix of the read.\r\n",
      "  -g ADAPTER, --front ADAPTER\r\n",
      "                        Sequence of an adapter ligated to the 5' end (paired\r\n",
      "                        data: of the first read). The adapter and any preceding\r\n",
      "                        bases are trimmed. Partial matches at the 5' end are\r\n",
      "                        allowed. If a '^' character is prepended ('anchoring'),\r\n",
      "                        the adapter is only found if it is a prefix of the read.\r\n",
      "  -b ADAPTER, --anywhere ADAPTER\r\n",
      "                        Sequence of an adapter that may be ligated to the 5' or\r\n",
      "                        3' end (paired data: of the first read). Both types of\r\n",
      "                        matches as described under -a and -g are allowed. If the\r\n",
      "                        first base of the read is part of the match, the\r\n",
      "                        behavior is as with -g, otherwise as with -a. This\r\n",
      "                        option is mostly for rescuing failed library\r\n",
      "                        preparations - do not use if you know which end your\r\n",
      "                        adapter was ligated to!\r\n",
      "  -e E, --error-rate E, --errors E\r\n",
      "                        Maximum allowed error rate (if 0 <= E < 1), or absolute\r\n",
      "                        number of errors for full-length adapter match (if E is\r\n",
      "                        an integer >= 1). Error rate = no. of errors divided by\r\n",
      "                        length of matching region. Default: 0.1 (10%)\r\n",
      "  --no-indels           Allow only mismatches in alignments. Default: allow both\r\n",
      "                        mismatches and indels\r\n",
      "  -n COUNT, --times COUNT\r\n",
      "                        Remove up to COUNT adapters from each read. Default: 1\r\n",
      "  -O MINLENGTH, --overlap MINLENGTH\r\n",
      "                        Require MINLENGTH overlap between read and adapter for\r\n",
      "                        an adapter to be found. Default: 3\r\n",
      "  --match-read-wildcards\r\n",
      "                        Interpret IUPAC wildcards in reads. Default: False\r\n",
      "  -N, --no-match-adapter-wildcards\r\n",
      "                        Do not interpret IUPAC wildcards in adapters.\r\n",
      "  --action {trim,retain,mask,lowercase,none}\r\n",
      "                        What to do if a match was found. trim: trim adapter and\r\n",
      "                        up- or downstream sequence; retain: trim, but retain\r\n",
      "                        adapter; mask: replace with 'N' characters; lowercase:\r\n",
      "                        convert to lowercase; none: leave unchanged. Default:\r\n",
      "                        trim\r\n",
      "  --rc, --revcomp       Check both the read and its reverse complement for\r\n",
      "                        adapter matches. If match is on reverse-complemented\r\n",
      "                        version, output that one. Default: check only read\r\n",
      "\r\n",
      "Additional read modifications:\r\n",
      "  -u LEN, --cut LEN     Remove LEN bases from each read (or R1 if paired; use -U\r\n",
      "                        option for R2). If LEN is positive, remove bases from\r\n",
      "                        the beginning. If LEN is negative, remove bases from the\r\n",
      "                        end. Can be used twice if LENs have different signs.\r\n",
      "                        Applied *before* adapter trimming.\r\n",
      "  --nextseq-trim 3'CUTOFF\r\n",
      "                        NextSeq-specific quality trimming (each read). Trims\r\n",
      "                        also dark cycles appearing as high-quality G bases.\r\n",
      "  -q [5'CUTOFF,]3'CUTOFF, --quality-cutoff [5'CUTOFF,]3'CUTOFF\r\n",
      "                        Trim low-quality bases from 5' and/or 3' ends of each\r\n",
      "                        read before adapter removal. Applied to both reads if\r\n",
      "                        data is paired. If one value is given, only the 3' end\r\n",
      "                        is trimmed. If two comma-separated cutoffs are given,\r\n",
      "                        the 5' end is trimmed with the first cutoff, the 3' end\r\n",
      "                        with the second.\r\n",
      "  --quality-base N      Assume that quality values in FASTQ are encoded as\r\n",
      "                        ascii(quality + N). This needs to be set to 64 for some\r\n",
      "                        old Illumina FASTQ files. Default: 33\r\n",
      "  --length LENGTH, -l LENGTH\r\n",
      "                        Shorten reads to LENGTH. Positive values remove bases at\r\n",
      "                        the end while negative ones remove bases at the\r\n",
      "                        beginning. This and the following modifications are\r\n",
      "                        applied after adapter trimming.\r\n",
      "  --trim-n              Trim N's on ends of reads.\r\n",
      "  --length-tag TAG      Search for TAG followed by a decimal number in the\r\n",
      "                        description field of the read. Replace the decimal\r\n",
      "                        number with the correct length of the trimmed read. For\r\n",
      "                        example, use --length-tag 'length=' to correct fields\r\n",
      "                        like 'length=123'.\r\n",
      "  --strip-suffix STRIP_SUFFIX\r\n",
      "                        Remove this suffix from read names if present. Can be\r\n",
      "                        given multiple times.\r\n",
      "  -x PREFIX, --prefix PREFIX\r\n",
      "                        Add this prefix to read names. Use {name} to insert the\r\n",
      "                        name of the matching adapter.\r\n",
      "  -y SUFFIX, --suffix SUFFIX\r\n",
      "                        Add this suffix to read names; can also include {name}\r\n",
      "  --rename TEMPLATE     Rename reads using TEMPLATE containing variables such as\r\n",
      "                        {id}, {adapter_name} etc. (see documentation)\r\n",
      "  --zero-cap, -z        Change negative quality values to zero.\r\n",
      "\r\n",
      "Filtering of processed reads:\r\n",
      "  Filters are applied after above read modifications. Paired-end reads are\r\n",
      "  always discarded pairwise (see also --pair-filter).\r\n",
      "\r\n",
      "  -m LEN[:LEN2], --minimum-length LEN[:LEN2]\r\n",
      "                        Discard reads shorter than LEN. Default: 0\r\n",
      "  -M LEN[:LEN2], --maximum-length LEN[:LEN2]\r\n",
      "                        Discard reads longer than LEN. Default: no limit\r\n",
      "  --max-n COUNT         Discard reads with more than COUNT 'N' bases. If COUNT\r\n",
      "                        is a number between 0 and 1, it is interpreted as a\r\n",
      "                        fraction of the read length.\r\n",
      "  --max-expected-errors ERRORS, --max-ee ERRORS\r\n",
      "                        Discard reads whose expected number of errors (computed\r\n",
      "                        from quality values) exceeds ERRORS.\r\n",
      "  --discard-trimmed, --discard\r\n",
      "                        Discard reads that contain an adapter. Use also -O to\r\n",
      "                        avoid discarding too many randomly matching reads.\r\n",
      "  --discard-untrimmed, --trimmed-only\r\n",
      "                        Discard reads that do not contain an adapter.\r\n",
      "  --discard-casava      Discard reads that did not pass CASAVA filtering (header\r\n",
      "                        has :Y:).\r\n",
      "\r\n",
      "Output:\r\n",
      "  --quiet               Print only error messages.\r\n",
      "  --report {full,minimal}\r\n",
      "                        Which type of report to print: 'full' or 'minimal'.\r\n",
      "                        Default: full\r\n",
      "  --json FILE           Dump report in JSON format to FILE\r\n",
      "  -o FILE, --output FILE\r\n",
      "                        Write trimmed reads to FILE. FASTQ or FASTA format is\r\n",
      "                        chosen depending on input. Summary report is sent to\r\n",
      "                        standard output. Use '{name}' for demultiplexing (see\r\n",
      "                        docs). Default: write to standard output\r\n",
      "  --fasta               Output FASTA to standard output even on FASTQ input.\r\n",
      "  -Z                    Use compression level 1 for gzipped output files\r\n",
      "                        (faster, but uses more space)\r\n",
      "  --info-file FILE      Write information about each read and its adapter\r\n",
      "                        matches into FILE. See the documentation for the file\r\n",
      "                        format.\r\n",
      "  -r FILE, --rest-file FILE\r\n",
      "                        When the adapter matches in the middle of a read, write\r\n",
      "                        the rest (after the adapter) to FILE.\r\n",
      "  --wildcard-file FILE  When the adapter has N wildcard bases, write adapter\r\n",
      "                        bases matching wildcard positions to FILE. (Inaccurate\r\n",
      "                        with indels.)\r\n",
      "  --too-short-output FILE\r\n",
      "                        Write reads that are too short (according to length\r\n",
      "                        specified by -m) to FILE. Default: discard reads\r\n",
      "  --too-long-output FILE\r\n",
      "                        Write reads that are too long (according to length\r\n",
      "                        specified by -M) to FILE. Default: discard reads\r\n",
      "  --untrimmed-output FILE\r\n",
      "                        Write reads that do not contain any adapter to FILE.\r\n",
      "                        Default: output to same file as trimmed reads\r\n",
      "\r\n",
      "Paired-end options:\r\n",
      "  The -A/-G/-B/-U/-Q options work like their lowercase counterparts, but are\r\n",
      "  applied to R2 (second read in pair)\r\n",
      "\r\n",
      "  -A ADAPTER            3' adapter to be removed from R2\r\n",
      "  -G ADAPTER            5' adapter to be removed from R2\r\n",
      "  -B ADAPTER            5'/3 adapter to be removed from R2\r\n",
      "  -U LENGTH             Remove LENGTH bases from R2\r\n",
      "  -Q [5'CUTOFF,]3'CUTOFF\r\n",
      "                        Quality-trimming cutoff for R2. Default: same as for R1\r\n",
      "  -p FILE, --paired-output FILE\r\n",
      "                        Write R2 to FILE.\r\n",
      "  --pair-adapters       Treat adapters given with -a/-A etc. as pairs. Either\r\n",
      "                        both or none are removed from each read pair.\r\n",
      "  --pair-filter {any,both,first}\r\n",
      "                        Which of the reads in a paired-end read have to match\r\n",
      "                        the filtering criterion in order for the pair to be\r\n",
      "                        filtered. Default: any\r\n",
      "  --interleaved         Read and/or write interleaved paired-end reads.\r\n",
      "  --untrimmed-paired-output FILE\r\n",
      "                        Write second read in a pair to this FILE when no adapter\r\n",
      "                        was found. Use with --untrimmed-output. Default: output\r\n",
      "                        to same file as trimmed reads\r\n",
      "  --too-short-paired-output FILE\r\n",
      "                        Write second read in a pair to this file if pair is too\r\n",
      "                        short.\r\n",
      "  --too-long-paired-output FILE\r\n",
      "                        Write second read in a pair to this file if pair is too\r\n",
      "                        long.\r\n"
     ]
    }
   ],
   "source": [
    "!cutadapt -h"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
