# DADA2 pipeline

__TEST__

Libraries

```{r}
.libPaths(c("/projappl/project_2007145/project_rpackages_r421", .libPaths()))
libpath <- .libPaths()[1]

library(dada2)
library(phyloseq)
library(microViz)
library(tidyverse)
```

## 16S rRNA data

First assing the path to the adta that has been trimmed from PCR primers.
```{r}
path <- "trimmed_data/bac_data"
```

Then we will make two lists of the read files in the folder, one for R1 and one for R2 read files.  
And get the samples names from the names of the read files (from R1 read files to be precise.)
```{r}
fnFs <- sort(list.files(path, pattern="_1_trimmed.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_2_trimmed.fastq.gz", full.names = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

To be able to decide the trimming parameters, we'll plot the quality profiles from first six read files. Separately for R1 and R2 reads.  

__Why? Any ideas?__

```{r}
plotQualityProfile(fnFs[1:6])
plotQualityProfile(fnRs[1:6])
````

Before stating to trim the reads, we define the output files for the trimmed reads.
```{r}
filtFs <- file.path(path, "filtered", paste0(sample.names, "_R1_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R2_filt.fastq.gz"))

names(filtFs) <- sample.names
names(filtRs) <- sample.names
````

Then we can quality trimn the reads, but first we need to decide some the trimming parameters based on the quality plots.  
Before running the trimming command, set the trimming length (`truncLen`) and the maximun expected errors per read. 
```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(XXX, XXX),
              maxN=0, maxEE=c(X,X), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE)

``` 

Then we can run the error learning function on the quality trimmed reads and plot the error profiles. 
```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)
```


```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE, pool=TRUE)
dadaRs <- dada(filtRs, err=errF, multithread=TRUE, pool=TRUE)

mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)

seqtab <- makeSequenceTable(mergers)
dim(seqtab)
table(nchar(getSequences(seqtab)))

seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)

track <- cbind(out, rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "nonchim")
rownames(track) <- sample.names
track

taxa <- assignTaxonomy(seqtab.nochim, "/scratch/project_2007145/databases/silva_nr99_v138.1_wSpecies_train_set.fa.gz",
                        minBoot=80, multithread=TRUE)

taxa.print <- taxa
rownames(taxa.print) <- NULL
head(taxa.print)

physeq <- phyloseq(otu_table(t(seqtab.nochim), taxa_are_rows=TRUE),
                 tax_table(taxa))

dna <- Biostrings::DNAStringSet(taxa_names(physeq))
names(dna) <- taxa_names(physeq)
physeq <- merge_phyloseq(physeq, dna)

taxa_names(physeq) <- paste0("ASV", seq(ntaxa(physeq)))

## remove unwanted (both missing from data)
# mitochondria
physeq <- prune_taxa(!(taxa_names(physeq) %in% taxa_names(subset_taxa(physeq, Family=="Mitochondria"))), physeq)
# chloroplasts
physeq <- prune_taxa(!(taxa_names(physeq) %in% taxa_names(subset_taxa(physeq, Order=="Chloroplast"))), physeq)

physeq_meta <- read_delim("doc/bac_meta.csv", delim=",", col_types="ccc") %>% column_to_rownames("Run")
sample_data(physeq) <- sample_data(physeq_meta)

saveRDS(physeq, "outputs/physeq_bac.rds")
```

## ITS data

```{r}
path <- "trimmed_data/fun_data"

fnFs <- sort(list.files(path, pattern="_1_trimmed.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_2_trimmed.fastq.gz", full.names = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)

plotQualityProfile(fnFs[1:12])
plotQualityProfile(fnRs[1:12])

filtFs <- file.path(path, "filtered", paste0(sample.names, "_R1_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R2_filt.fastq.gz"))

names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs,
              maxN=0, maxEE=c(2,3), truncQ=2, rm.phix=TRUE, minLen=75,
              compress=TRUE, multithread=TRUE)

errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)

dadaFs <- dada(filtFs, err=errF, multithread=TRUE, pool=TRUE)
dadaRs <- dada(filtRs, err=errF, multithread=TRUE, pool=TRUE)

mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)

seqtab <- makeSequenceTable(mergers)
dim(seqtab)
table(nchar(getSequences(seqtab)))

seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)

track <- cbind(out, rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "nonchim")
rownames(track) <- sample.names
track

taxa <- assignTaxonomy(seqtab.nochim, "/scratch/project_2007145/databases/sh_general_release_dynamic_29.11.2022.fasta",
                        minBoot=80, multithread=TRUE, tryRC = TRUE)

# for pretty printing of the taxonomy
taxa.print <- taxa
rownames(taxa.print) <- NULL
head(taxa.print)

physeq <- phyloseq(otu_table(t(seqtab.nochim), taxa_are_rows=TRUE),
                 tax_table(taxa))

dna <- Biostrings::DNAStringSet(taxa_names(physeq))
names(dna) <- taxa_names(physeq)
physeq <- merge_phyloseq(physeq, dna)

taxa_names(physeq) <- paste0("ASV", seq(ntaxa(physeq)))

physeq_meta <- read_delim("doc/fun_meta.csv", delim=",", col_types="ccc") %>% column_to_rownames("Run")
sample_data(physeq) <- sample_data(physeq_meta)

saveRDS(physeq, "outputs/physeq_fun.rds")
```